exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -448,39 +433,64 @@ public Builder withValidators(short validatorCount) {
     }
 
     /**
-     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     * Adds a service artifact which would be deployed by the TestKit. Several service artifacts
+     * can be added.
+     *
+     * <p>Once the service artifact is deployed, the service instances can be added with
+     * {@link #withService(ServiceArtifactId, String, int, MessageLite)}.
      */
-    public Builder withService(Class<? extends ServiceModule> serviceModule) {
-      services.add(serviceModule);
+    public Builder withDeployedArtifact(ServiceArtifactId serviceArtifactId, String artifactFilename) {
+      serviceArtifactFilenames.put(serviceArtifactId, artifactFilename);
       return this;
     }
 
     /**
-     * Adds services with which the TestKit would be instantiated.
+     * Adds a service specification with which the TestKit would create the corresponding service
+     * instance. Several service specifications can be added. All services are started and
+     * configured before the genesis block.
+     *
+     * <p>Note that the corresponding service artifact with equal serviceArtifactId should be
+     * deployed with {@link #withDeployedArtifact(ServiceArtifactId, String)}.
      */
-    @SafeVarargs
-    public final Builder withServices(Class<? extends ServiceModule> serviceModule,
-                                      Class<? extends ServiceModule>... serviceModules) {
-      return withServices(Lists.asList(serviceModule, serviceModules));
+    public Builder withService(ServiceArtifactId serviceArtifactId, String serviceName,
+                               int serviceId, MessageLite configuration) {
+      checkServiceId(serviceId);
```

Comment:
@skletsun 
Does testkit/core check properly (with reasonable error messages)
that there are no duplicate ids assigned? The builder accepts ids for regular (here) and time services, and something has to check they do not intersect.

Reply:
> @skletsun
> Does testkit/core check properly (with reasonable error messages)
> that there are no duplicate ids assigned? The builder accepts ids for regular (here) and time services, and something has to check they do not intersect.

Yes, it's checked in the dispatcher code while starting service.